libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,467 @@
+        for (uint32_t i=0; i<_database.count && !found; i++) {
+            // skip last hit because we already checked it above
+            if (i != _database.last_hit && is_item_in_database(i, item)) {
+                found = true;
+                _database.last_hit = i;
+            }
+        }
+
+        // if we got here we should add to database
+        if (!found) {
+            database_item_add(item);
+        } else {
+            _queue.stats.dropped++;
+        }
+    }
+
+    database_items_remove_all_expired();
+}
+
+// returns true if a similar object already exists in database. When true, the object timer is also reset
+bool AP_OADatabase::is_item_in_database(const uint32_t index, const OA_DbItem &item)
+{
+    if (index >= _database.count) {
+        // index out of range
+        return false;
+    }
+    if (_database.items[index].loc.lat == item.loc.lat &&
+        //_database.items[index].loc.alt == obj.loc.alt &&
+        _database.items[index].loc.lng == item.loc.lng)
+    {
+        // exact match
+        database_item_refresh(index, item.timestamp);
+        return true;
+    }
+
+    const float dist = _database.items[index].loc.get_distance(item.loc);
+    if (dist < _database.filter_m) {
+        // update timestamp on nearby object so it sticks around longer
+        database_item_refresh(index, item.timestamp);
+        return true;
+    }
+    return false;
+}
+
+void AP_OADatabase::database_items_remove_all_expired()
+{
+    // calculate age of all items in the _database
+
+    if (_database.expiry_seconds == 0) {
+        // zero means never expire. This is not normal behavior but perhaps you could send a static
+        // environment once that you don't want to have to constantly update
+        return;
+
+    } else if (_database.expiry_seconds < 0) {
+        // invalid param, reset to default
+        _database.expiry_seconds.set_and_notify(AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT);
+    }
+
+    const uint32_t now_ms = AP_HAL::millis();
+    const uint32_t expiry_ms = (uint32_t)_database.expiry_seconds * 1000;
+    uint32_t index = 0;
+    while (index < _database.count) {
+        if (now_ms - _database.items[index].timestamp > expiry_ms) {
+            database_item_remove(index);
+        } else {
+            index++;
+        }
+    }
+}
+
+void AP_OADatabase::database_item_remove(const uint32_t index)
+{
+    //::printf("DB::database_item_remove()\n");
+
+    if (index >= _database.count || _database.count == 0) {
+        // index out of range
+        return;
+    }
+
+    _database.stats.dropped++;
+
+    // invalidate the location so that a UI will redraw it
+    _database.items[index] = _database.invalid_item;
+
+    database_item_send(index);
+
+    _database.count--;
+    if (_database.count == 0) {
+        return;
+    }
+
+    if (index != _database.count) {
+        // copy last object in array over expired object
+        _database.items[index] = _database.items[_database.count];
```

Comment:
Presumably if last_hit was the last index you should also be updating here on this copy.